File: selected_coding_interview/codes/python/lc_79_word_search.py
"""
File: lc_79_word_search.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(i, j, k):
            if not 0 <= i < len(board) or not 0 <= j < len(board[0]) or board[i][j] != word[k]: return False
            if k == len(word) - 1: return True
            board[i][j] = ''
            res = dfs(i + 1, j, k + 1) or dfs(i - 1, j, k + 1) or dfs(i, j + 1, k + 1) or dfs(i, j - 1, k + 1)
            board[i][j] = word[k]
            return res
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, 0): return True
        return False


# ======= Test Case =======
test_input_board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]
test_input_word = "ABCCED"
expected_output = True

# ====== Driver Code ======
slt = Solution()
result = slt.exist(test_input_board, test_input_word)
print(result)

File: selected_coding_interview/codes/java/lc_79_word_search/lc_79_word_search.java
/*
* File: lc_79_word_search.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_79_word_search;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public boolean exist(char[][] board, String word) {
            char[] words = word.toCharArray();
            for(int i = 0; i < board.length; i++) {
                for(int j = 0; j < board[0].length; j++) {
                    if (dfs(board, words, i, j, 0)) return true;
                }
            }
            return false;
        }
        boolean dfs(char[][] board, char[] word, int i, int j, int k) {
            if (i >= board.length || i < 0 || j >= board[0].length || j < 0 || board[i][j] != word[k]) return false;
            if (k == word.length - 1) return true;
            board[i][j] = '\0';
            boolean res = dfs(board, word, i + 1, j, k + 1) || dfs(board, word, i - 1, j, k + 1) || 
                          dfs(board, word, i, j + 1, k + 1) || dfs(board, word, i , j - 1, k + 1);
            board[i][j] = word[k];
            return res;
        }
    }

public class lc_79_word_search {
    public static void main(String[] args) {
        // ======= Test Case =======
        String[][] test_input_board = new String[][]{{"A","B","C","E"}, {"S","F","C","S"}, {"A","D","E","E"}};
        String test_input_word = "ABCCED";
        var expected_output = true;

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.exist(test_input_board, test_input_word);
        System.out.println(result);

    }
}
